tskv/src/kvcore.rs

```diff
@@ -12,39 +29,87 @@ use once_cell::sync::OnceCell;
 use std::sync::Arc;
 use std::thread::JoinHandle;
 
+pub struct Entry {
+    pub series_id: u64,
+}
+
 #[derive(Clone)]
 pub struct TsKv {
+    options: Options,
     kvctx: Arc<KvContext>,
     version_set: Arc<VersionSet>,
-}
+    file_manager: Arc<FileManager>,
 
-pub struct Entry {
-    pub series_id: u64,
+    pool: Arc<ThreadPool>,
+    wal_sender: UnboundedSender<WalTask>,
 }
 
 impl TsKv {
-    pub fn open(opt: Options) -> Self {
-        let kvctx = Arc::new(KvContext::new(opt));
+    pub fn open(opt: Options) -> Result<Self> {
+        let kvctx = Arc::new(KvContext::new(opt.clone()));
         let vs = Self::recover();
-        Self {
+
+        // let file_manager = FileManager::get_instance();
+        let (sender, receiver) = unbounded();
+
+        let core = Self {
+            options: opt,
             kvctx,
             version_set: Arc::new(vs),
-        }
+            file_manager: Arc::new(FileManager::new()),
+            pool: Arc::new(ThreadPool::new().unwrap()),
+            wal_sender: sender,
+        };
+
+        core.run_wal_job(receiver);
+
+        Ok(core)
     }
+
     pub fn recover() -> VersionSet {
         //todo! recover from manifest and build VersionSet
         VersionSet::new_default()
     }
+
     pub fn get_instance() -> &'static Self {
         static INSTANCE: OnceCell<TsKv> = OnceCell::new();
-        INSTANCE.get_or_init(|| Self::open(Options::from_env()))
+        INSTANCE.get_or_init(|| Self::open(Options::from_env()).unwrap())
     }
-    pub async fn write(&self, write_batch: Vec<Entry>) -> Result<()> {
-        //wal
-        let _ = self.kvctx.shard_write(0, write_batch).await;
+
+    pub async fn write(&mut self, write_batch: WriteRowsRpcRequest) -> Result<()> {
+        let (cb, rx) = oneshot::channel();
+        self.wal_sender
+            .send(WalTask::Write {
+                rows: write_batch.rows,
+                cb,
+            })
+            .await
+            .map_err(|err| Error::Send)?;
+        rx.await.unwrap().unwrap();
+
+        // let _ = self.kvctx.shard_write(0, write_batch).await;
         Ok(())
     }
 
+    pub fn run_wal_job(&self, mut receiver: UnboundedReceiver<WalTask>) {
+        let fm = Arc::clone(&self.file_manager);
+        let wal_opt = self.options.wal.clone();
+        let f = async move {
+            let mut wal_manager = WalFileManager::new(fm, wal_opt);
+            while let Some(x) = receiver.next().await {
+                match x {
+                    WalTask::Write { rows, cb } => {
+                        let ret = wal_manager.write(&rows).await;
```

Comment:
result allways ok ?

Reply:
will be fixed on next commit

Reply:
fixed in [09cf1f0](https://github.com/cnosdb/cnosdb/pull/237/commits/09cf1f08d3319e364443f8f512ac9a1ccdc77550)